lib/common/ipc_client.c

```diff
@@ -31,6 +31,672 @@
+    }
+}
+
+/*!
+ * \internal
+ * \brief Dispatch data read from IPC source
+ *
+ * \param[in] buffer     Data read from IPC
+ * \param[in] length     Number of bytes of data in buffer (ignored)
+ * \param[in] user_data  IPC object
+ *
+ * \return Always 0 (meaning connection is still required)
+ *
+ * \note This function can be used as a main loop IPC dispatch callback.
+ */
+static int
+dispatch_ipc_data(const char *buffer, ssize_t length, gpointer user_data)
+{
+    pcmk_ipc_api_t *api = user_data;
+    xmlNode *msg;
+
+    CRM_CHECK(api != NULL, return 0);
+
+    if (buffer == NULL) {
+        crm_warn("Empty message received from %s IPC",
+                 pcmk_ipc_name(api, true));
+        return 0;
+    }
+
+    msg = string2xml(buffer);
+    if (msg == NULL) {
+        crm_warn("Malformed message received from %s IPC",
+                 pcmk_ipc_name(api, true));
+        return 0;
+    }
+    call_api_dispatch(api, msg);
+    free_xml(msg);
+    return 0;
+}
+
+/*!
+ * \brief Check whether an IPC connection has data available (without main loop)
+ *
+ * \param[in]  api         IPC API connection
+ * \param[in]  timeout_ms  If less than 0, poll indefinitely; if 0, poll once
+ *                         and return immediately; otherwise, poll for up to
+ *                         this many milliseconds
+ *
+ * \return Standard Pacemaker return code
+ *
+ * \note Callers of pcmk_connect_ipc() using pcmk_ipc_dispatch_poll should call
+ *       this function to check whether IPC data is available. Return values of
+ *       interest include pcmk_rc_ok meaning data is available, and EAGAIN
+ *       meaning no data is available; all other values indicate errors.
+ * \todo This does not allow the caller to poll multiple file descriptors at
+ *       once. If there is demand for that, we could add a wrapper for
+ *       crm_ipc_get_fd(api->ipc), so the caller can call poll() themselves.
+ */
+int
+pcmk_poll_ipc(pcmk_ipc_api_t *api, int timeout_ms)
+{
+    int rc;
+    struct pollfd pollfd = { 0, };
+
+    if ((api == NULL) || (api->dispatch_type != pcmk_ipc_dispatch_poll)) {
+        return EINVAL;
+    }
+    pollfd.fd = crm_ipc_get_fd(api->ipc);
+    pollfd.events = POLLIN;
+    rc = poll(&pollfd, 1, timeout_ms);
+    if (rc < 0) {
+        return errno;
+    } else if (rc == 0) {
+        return EAGAIN;
+    }
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \brief Dispatch available messages on an IPC connection (without main loop)
+ *
+ * \param[in]  api  IPC API connection
+ *
+ * \return Standard Pacemaker return code
+ *
+ * \note Callers of pcmk_connect_ipc() using pcmk_ipc_dispatch_poll should call
+ *       this function when IPC data is available.
+ */
+void
+pcmk_dispatch_ipc(pcmk_ipc_api_t *api)
+{
+    if (api != NULL) {
+        while (crm_ipc_ready(api->ipc)) {
+            if (crm_ipc_read(api->ipc) > 0) {
+                dispatch_ipc_data(crm_ipc_buffer(api->ipc), 0, api);
+            }
+        }
+    }
+}
```

Comment:
Just a stylistic thing here... I like to invert the condition to save the extra layer of indentation.  So:
```
    if (api == NULL) {
        return;
    }
    while (crm_ipc_ready(api->ipc)) {
    ....
```

Reply:
makes sense